File: Pythoncode/44.通配符匹配.py
class Solution:
    def isMatch(self, s, p):
        p = self.deldupstar(p)
        dp = [[None for _ in range(len(p)+1)] for _ in range(len(s)+1)]
        for j in reversed(range(len(p)+1)):
            for i in reversed(range(len(s)+1)):
                if i == len(s):
                    self.prev = False
                self.prev = self.helper(s, p, dp, i, j) or self.prev
                if j - 1 >= 0 and p[j-1] == "*":
                    dp[i][j-1] = self.prev
        return dp[0][0]

    def deldupstar(self, p):
        i = 0
        j = 0
        p = list(p)
        while j < len(p):
            if p[j] != "*":
                p[i] = p[j]
                i += 1
                j += 1
            else:
                if j+1 >= len(p) or p[j+1] != "*":
                    p[i] = p[j]
                    i += 1
                    j += 1
                else:
                    j += 1
        return "".join(p[:i])

    def helper(self, s, p, dp, s_idx, p_idx):
        if dp[s_idx][p_idx] != None:
            return dp[s_idx][p_idx]
        if len(p) == p_idx and len(s) == s_idx:
            dp[s_idx][p_idx] = True
            return dp[s_idx][p_idx]
        if len(p) == p_idx and len(s) != s_idx:
            dp[s_idx][p_idx] = False
            return dp[s_idx][p_idx]
        if len(p) != p_idx and len(s) == s_idx:
            if p[p_idx] == "*":
                dp[s_idx][p_idx] = dp[s_idx][p_idx+1]
            else:
                dp[s_idx][p_idx] = False
            return dp[s_idx][p_idx]
        
        if p[p_idx] in [s[s_idx], "?"]:
            dp[s_idx][p_idx] = dp[s_idx+1][p_idx+1]
        else:
            dp[s_idx][p_idx] = False
        return dp[s_idx][p_idx]

File: Cppcode/44.通配符匹配.cpp
#include <string>
#include <vector>

using namespace std;
class Solution {
public:
    vector<vector<int>> dp;
    bool isMatch(string s, string p) {
        dp = vector<vector<int>> (s.size()+1, vector<int>(p.size()+1, 0));
        for (decltype(s.size()) i = 0; i <= s.size(); i++)
        {
            for (decltype(p.size()) j = 0; j <= p.size(); j++)
            {
                helper(s.substr(0, i), p.substr(0, j));
            }
        }
        if (dp[s.size()][p.size()] == 1) return true;
        else return false;
    }
    void helper(string s, string p)
    {
        if (p.size() == 0 and s.size() == 0) dp[s.size()][p.size()] =  1;
        else if (p.size() == 0 and s.size() != 0) dp[s.size()][p.size()] = -1;
        else
        {
            if (p.back() != '*')
            {
                if (s.size() == 0 or (p.back() != s.back() and p.back() != '?')) dp[s.size()][p.size()] = -1;
                else dp[s.size()][p.size()] = dp[s.size()-1][p.size()-1];
            }
            else
            {
                if (dp[s.size()][p.size()-1] == 1) dp[s.size()][p.size()] =  1; // 该‘*’不匹配任何字符
                else if (s.size() > 0) dp[s.size()][p.size()] =  dp[s.size()-1][p.size()]; // 该'*'至少匹配了一个字符
                else dp[s.size()][p.size()] = -1;
            }
        }
    }
};

// int main()
// {
//     Solution s;
//     s.isMatch("aa","*");
// }